### PR TITLE
initial commit for setup list

### PR DESCRIPTION
**Pull request by janvanrijn**
added a function to obtain setup lists

### CODE REVIEW

File: openml/flows/sklearn_converter.py [hunk 1]
```diff
@@ -533,7 +533,7 @@ def _serialize_cross_validator(o):
     for key in args:
         # We need deprecation warnings to always be on in order to
         # catch deprecated param values.
-        # This is set in utils/__init__.py but it gets overwritten
+        # This is set in utils/__init__.py.py.py but it gets overwritten
```
Comment (mfeurer) [inline openml/flows/sklearn_converter.py]: Why do you add `.py.py`?

File: openml/setups/functions.py [hunk 2]
```diff
@@ -74,6 +74,76 @@ def get_setup(setup_id):
     return _create_setup_from_xml(result_dict)
 
 
+def setup_list(flow=None, tag=None, offset=None, size=None):
```
Comment (mfeurer) [inline openml/setups/functions.py]: Wouldn't `list_setups` be a better name?

File: openml/setups/functions.py [hunk 3]
```diff
@@ -74,6 +74,76 @@ def get_setup(setup_id):
...
+    xml_string = openml._api_calls._perform_api_call(api_call)
+
+    setups_dict = xmltodict.parse(xml_string)
+    # Minimalistic check if the XML is useful
+    if 'oml:setups' not in setups_dict:
+        raise ValueError('Error in return XML, does not contain "oml:setups": %s'
+                         % str(setups_dict))
+    elif '@xmlns:oml' not in setups_dict['oml:setups']:
+        raise ValueError('Error in return XML, does not contain '
+                         '"oml:runs"/@xmlns:oml: %s'
```
Comment (mfeurer) [inline openml/setups/functions.py]: The error message looks like there is a copy-paste error as it references runs.

File: openml/setups/functions.py [hunk 4]
```diff
@@ -74,6 +74,76 @@ def get_setup(setup_id):
...
+                         % str(setups_dict))
+    elif setups_dict['oml:setups']['@xmlns:oml'] != 'http://openml.org/openml':
+        raise ValueError('Error in return XML, value of  '
+                         '"oml:runs"/@xmlns:oml is not '
```
Comment (mfeurer) [inline openml/setups/functions.py]: Same here.

File: openml/setups/functions.py [hunk 5]
```diff
@@ -74,6 +74,76 @@ def get_setup(setup_id):
...
+        setups_list = setups_dict['oml:setups']['oml:setup']
+    elif isinstance(setups_dict['oml:setups']['oml:setup'], dict):
+        setups_list = [setups_dict['oml:setups']['oml:setup']]
+    else:
+        raise TypeError()
+
+    setups = dict()
+    for setup_ in setups_list:
+        # making it a dict to give it the right format
+        current = _create_setup_from_xml({'oml:setup_parameters': setup_})
```
Comment (mfeurer) [inline openml/setups/functions.py]: This seems to be inconsistent with the other list functions which only return a describing dictionary. Is this already like this on the API side?

File: openml/study/functions.py [hunk 6]
```diff
@@ -27,9 +27,11 @@ def get_study(study_id):
     creator = result_dict['oml:creator']
     tags = []
     for tag in _multitag_to_list(result_dict, 'oml:tag'):
-        tags.append({'name': tag['oml:name'],
-                     'window_start': tag['oml:window_start'],
-                     'write_access': tag['oml:write_access']})
+        current_tag = {'name': tag['oml:name'],
+                       'write_access': tag['oml:write_access']}
```
Comment (mfeurer) [inline openml/study/functions.py]: Hm, what is `write_access` and `window_start`?

File: openml/setups/functions.py (hunk 5 above)
Comment (janvanrijn) [inline openml/setups/functions.py]: In what sense inconsitent?

File: openml/study/functions.py (hunk 6 above)
Comment (janvanrijn) [inline openml/study/functions.py]: the tags that are associated with a study determine which entities belong to the study. window start and write_access are both specifying this in more detail

**Comment by janvanrijn** [comment]
Agreed on almost all points , now fixed

File: openml/setups/functions.py (hunk 5 above)
Comment (mfeurer) [inline openml/setups/functions.py]: Inconsistent in the sense that other list functions return a dictionary mapping from the ID to another dictionary describing the dataset/task etc. Here, the list function returns a dict which maps from ID to a setup object. (Check [here](https://github.com/openml/openml-python/blob/master/openml/datasets/functions.py#L138) for the dataset list function.)

**Comment by mfeurer** [comment]
Hm, it seems like there's something messed up with the imports (looks like some form of cyclic import).

File: openml/setups/functions.py (hunk 5 above)
Comment (janvanrijn) [inline openml/setups/functions.py]: I see your point, but in this sense this is more rich, right? 

This listing function returns all information available, whereas other functions show just a summary of the function. There is no point in throwing this away

Comment (mfeurer) [inline openml/setups/functions.py]: Okay, makes sense.

File: openml/study/functions.py (hunk 6 above)
Comment (mfeurer) [inline openml/study/functions.py]: Hm, is this documented somewhere? I never saw this before and still don't really know what this additional information is good for.

**Pull request merged**